src/thunderbots/software/geom/util.h

```diff
@@ -338,6 +338,42 @@ std::vector<Point> lineIntersection(const Segment &a, const Segment &b);
  */
 Point reflect(const Point &v, const Point &n);
 
+/**
+ * Calculates the intersection point (if it exists) of an moving element with a line
+ * defined by 2 Points
+ *
+ * @param element_velocity: X/Y velocity of the element, defined in meters per second
+ * @param element_position: X/Y position of the element, defined in meters
+ * @param line_start: X/Y point where the line begins, defined in meters
+ * @param line_end: X/Y point where the line ends, defined in meters
+ * @param boundry_area: Defines the rectangular geometry that limits the scope of
+ * calculating an intersection. Only intersections within boundry_area will be considered,
+ * and the element position must exist within the boundry.
+ */
+std::optional<Point> velocityLineIntersection(const Vector element_velocity,
+                                              const Vector element_position,
+                                              const Point line_start,
+                                              const Point line_end,
+                                              const Rectangle boundry_area);
```

Comment:
A few points here:
- I think you're looking for the concept of a `Ray` (ie. a vector with a start point)
- We should probably take in a `Segment` here, rather then two points
- Why do we have the boundary area?
- What about line segments that overlap the ray?

I took a quick stab at a revised function signature, let me know what you think:
```
/**
 * Finds the intersection of a given Ray and Segment, if there is one
 *
 * @param ray
 * @param segment
 *
 * @return
 *      - a tuple of (std::nullopt, std::nullopt) if there is no intersection point
 *      - a tuple of (point, std::nullopt) if there is one intersection point
 *      - a tuple of (point, point), if the ray and segment are parallel, where the two
 *        points are the end-points of the line segment made up by the overlap of the
 *        ray and line segment
 */
std::tuple<std::optional<Point>, std::optional<Point>> raySegmentIntersect(const Ray& ray, const Segment& segment);
```

Reply:
Yeah this is a good idea. I implemented this function now.